client/post-editor/post-editor.jsx

```diff
@@ -1054,17 +1056,254 @@ export const PostEditor = React.createClass( {
+			const tagContent = content.substr( lastLtPos );
+			const tagMatch = tagContent.match( /<\s*(\/)?(\w+)/ );
+			if ( ! tagMatch ) {
+				return null;
+			}
+
+			const tagType = tagMatch[ 2 ];
+			const closingGt = tagContent.indexOf( '>' );
+			const isClosingTag = !! tagMatch[ 1 ];
+
+			return {
+				ltPos: lastLtPos,
+				gtPos: lastLtPos + closingGt + 1, // offset by one to get the position _after_ the character,
+				tagType,
+				isClosingTag,
+			};
+		}
+		return null;
+	},
+
+	getCursorMarkerSpan: function( type ) {
+		const tagType = type ? type : 'start';
+
+		return '<span ' +
+			'data-mce-type="bookmark"	' +
+			'id="mce_SELREST_' + tagType + '" ' +
+			'data-mce-style="overflow:hidden;line-height:0px" ' +
+			'style="overflow:hidden;line-height:0px"' +
+			'>' +
+			'&#65279;' +
+			'</span>';
+	},
+
+	addHTMLBookmarkInTextAreaContent: function() {
+		const textArea = this.editor._editor.getElement();
+
+		let htmlModeCursorStartPosition = textArea.selectionStart;
+		let htmlModeCursorEndPosition = textArea.selectionEnd;
+
+		// check if the cursor is in a tag and if so, adjust it
+		const isCursorStartInTag = this.getContainingTagInfo( textArea.value, htmlModeCursorStartPosition );
+		if ( isCursorStartInTag ) {
+			htmlModeCursorStartPosition = isCursorStartInTag.ltPos;
+		}
+
+		const isCursorEndInTag = this.getContainingTagInfo( textArea.value, htmlModeCursorEndPosition );
+		if ( isCursorEndInTag ) {
+			htmlModeCursorEndPosition = isCursorEndInTag.gtPos;
+		}
+
+		const mode =
+			htmlModeCursorStartPosition !== htmlModeCursorEndPosition
+			? 'range'
+			: 'single';
+
+		let selectedText = null;
+
+		if ( mode === 'range' ) {
+			const bookMarkEnd = this.getCursorMarkerSpan( 'end' );
+
+			selectedText = [
+				textArea.value.slice( htmlModeCursorStartPosition, htmlModeCursorEndPosition ),
+				bookMarkEnd,
+			].join( '' );
+		}
+
+		textArea.value = [
+			textArea.value.slice( 0, htmlModeCursorStartPosition ), // text until the cursor/selection position
+			this.getCursorMarkerSpan( 'start' ), 					// cursor/selection start marker
+			selectedText, 											// selected text with end cursor/position marker
+			textArea.value.slice( htmlModeCursorEndPosition )		// text from last cursor/selection position to end
+		].join( '' );
+
+		this.editor.onTextAreaChange( { target: { value: textArea.value } } );
+	},
+
+	focusHTMLBookmarkInVisualEditor: function( ed ) {
+		const startNode = ed.target.getDoc().getElementById( 'mce_SELREST_start' );
+		const endNode = ed.target.getDoc().getElementById( 'mce_SELREST_end' );
+
+		if ( ! startNode ) {
+			return;
+		}
+
+		ed.target.focus();
+
+		if ( ! endNode ) {
+			ed.target.selection.select( startNode );
+		} else {
+			const selection = document.createRange();
+			selection.setStart( startNode, 0 );
+			selection.setEnd( endNode, 0 );
+
+			ed.target.selection.setRng( selection );
+			endNode.parentNode.removeChild( endNode );
+		}
+
+		// TODO: Scroll to cursor position?
```

Comment:
Can we pull this TODO out of a comment and into an issue before merging, please?

I think this behavior would be desirable.